### PR TITLE
Implement dominace tree building algorithm

### PR DESCRIPTION
**Pull request by akiramenai**
Muchnick's implementation of this algo is insane (see p. 184 of the book). I reused previously calculated dominators instead.

P.S. Should I write it in the commit message?

### CODE REVIEW

**Comment by akiramenai** [comment]
Assume that variant works with references. Boost does and we are going to disable std version for now. I'm looking for a better solution for future.

File: src/Analysis/dominance.cpp [hunk 1]
```diff
@@ -2,22 +2,51 @@
...
+  if (LHS.count(Elem) != 0u) {
+    LHS.erase(Elem);
+    return true;
+  }
+  return false;
+}
+
+template <typename T>
+std::unordered_set<T> operator-(std::unordered_set<T> LHS, T Elem) {
+  subtractSets(LHS, Elem);
```
Comment (skyeycat) [inline src/Analysis/dominance.cpp]: Looks like you do not need a separate function to erase an element. Just erase it.

File: include/graph.h [hunk 2]
```diff
@@ -30,12 +30,21 @@ class Graph {
   // \brief Graph nodes listed in DFS order.
   std::vector<std::size_t> DFSOrder() const;
 
+  /// \return If the graph has \p arc
+  bool hasArc(Arc arc) const {
```
Comment (skyeycat) [inline include/graph.h]: If we use LLVM coding standard, you need to do smth with 'arc' (first upper case letter).

**Review by PavelKopyl — COMMENTED** [review]
LGTM

**Review by skyeycat — APPROVED** [review]
LGTM

**Review by PavelKopyl — APPROVED** [review]

**Pull request merged**